---
title: "Code_and_Workflow"
author: "Laura Tibbs"
date: "October 3, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part I: Replicate Unix assignment in R

Before beginning, load the tidyverse package:
```{r}
library(tidyverse)
```

First, set up the project with the following format: 
- `BCB546X_R_assignment/src` folder contains this code as well as the final R Markdown file (including graphs) and is set as the working directory (using `setwd()`)
- `BCB546X_R_assignment/data` folder already contains the input data files
- `BCB546X_R_assignment/output` folder is initially empty but will contain output files

If the input data is not already in the `data` folder, run the following to download the data:
```{r} 
getwd() #check that the working directory ends in "BCB546X_R_assignment/src"

# Download files:
download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", destfile = "../data/fang_et_al_genotypes.txt")
download.file("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/snp_position.txt", destfile="../data/snp_position.txt")
```
### Data Inspection

First, we need to read data from the files:

```{r}
# Load data:
genotypes <-  read_tsv("../data/fang_et_al_genotypes.txt")
snps <- read_tsv("../data/snp_position.txt")
```

Next, check the "Parsed with column specification..." message in the console to be sure that R guessed the data types correctly:

The `genotypes` file seems to have the correct column types.
However, it initially appears that not all of the columns in`snps` do. `Chromosome` and `Position` should be integer rather than character columns. However, first check that this is in fact the case by looking at all values of these columns:

```{r}
unique(snps$Chromosome)
# unique(snps$Position) # this shows that "multiple" is in this column, but prints a very long output to the file.
head(unique(snps$Position)) # use `head()` here to show only the first part of the long list of positions as an example.
```

This shows that there are in fact entries in these columns that are not integers, specifically "multiple" and "unknown". So, it is correct that these are not parsed as integers. If they really should have been integers, we could fix this by reading the data in again but manually specifying these column types using the `col_types` of `read_csv()`.

Now, we can inspect the data:

```{r}
genotypes
snps
```
Because `genotypes` and `snps` ae tibbles, R will automatically print only a manageable amount to the screen, even if I don't use `head()` to show only the first part of the data. From what is printed to the screen, I can see that `genotypes` is a tibble with 2782 rows and 986 columns, while `snps` is a tibble with 983 rows and 15 columns. The column names are also visible in this output.

If these objects were not tibbles, we could instead find the type and dimensions as follows:
``` {r}
typeof(genotypes)
typeof(snps)

dim(genotypes)
dim(snps)
```
These give the same dimensions as already listed above, and tells that the data frames are both of type "list". This is expected because tibbles (and data frames) are really just a specific type of list.

The tibbles' overall structure and their attributes (names, dimensions, and class) can also be confirmed with `str()` and `attributes()` (the output of this is quite long, so I will not print the output to the Rmd file, only the code used): 

``` {r eval=FALSE}
str(snps)
str(genotypes)

attributes(snps)
attributes(genotypes)
```

Alternately, the name attributes can be found using `names()` (here, I use `head(names(genotypes))` to show only the first few names, but they could all be viewed by removing `head()`):
```{r}
names(snps)
head(names(genotypes))
```

Finally, the size of the data as R objects can be found using `object.size()`:
```{r}
object.size(snps)
object.size(genotypes)
```
Which shows that the sizes of `snps` and `genotypes` are `r object.size(snps)` and `r object.size(genotypes)`, respectively.

### Data Processing


In order to join the `genotypes` and `snps` data to have a file with SNP_ID in the first column, "Chromosome" in the next, "Position" in the third, and individuals' genotype data in the other columns, the `genotype` data frame first needs to be transposed:

```{r}
tr.genotypes <- t(genotypes) # transpose
str(tr.genotypes) # check the structure of the new dataframe
typeof(tr.genotypes)
```

This transposed the file, but also made it into a character matrix rather than a data frame (a type of list) with columns of different types. To make the data tidier to work with, we can convert the matrix to the tidyverse's tibble data structure:

```{r}
tr.genotypes <- as.tibble(tr.genotypes)
typeof(tr.genotypes)
head(tr.genotypes)
```
Converting to a tibble makes it easy to see the types of the columns and the first few rows of their content. We can also see that the columns have been renamed V1, etc. However, we want the column names to be the individual IDs, so fix this using `colnames()`:

```{r}
colnames(tr.genotypes) <- tr.genotypes[1,] #set the column names to the values currently in the first row
head(tr.genotypes)
```

Next, look at the row names. These should match the column names from the original data because the data was transposed, but currently this is not the case. We can fix this manually, but setting row names on tibbles in tidyverse is deprecated, so instead add a new column that contains these values:
```{r}
head(rownames(tr.genotypes)) # these do not match the expected row names
tr.genotypes <- add_column(tr.genotypes, SNP_ID = colnames(genotypes), .before=1) # manually make a new column with "row names"
```

Now, join the transposed genotype file with the snp info file and check it for problems:
``` {r}
joined.data <- left_join(tr.genotypes, snps, by="SNP_ID") %>% #join the data
                select(SNP_ID, Chromosome, Position,everything())# and move the position columns to the front
head(joined.data) # check the new file
head(anti_join(snps, tr.genotypes, by="SNP_ID")) # use anti_join to check for unexpected ways that the match might have failed
head(anti_join(tr.genotypes, snps, by="SNP_ID"))
```
Because the only rows found using `anti_join()` were Sample_ID, JG_OTU, and Group, which were known to be absent from the `snps` file, and because the head of the new file looks as it should, the join appears to be successful.

Now, tidy the joined file by removing unnecessary rows and columns:
```{r}
joined.data <- filter(joined.data, !((SNP_ID == "Sample_ID") | (SNP_ID=="JG_OTU"))) %>% #remove unneeded Sample_ID and JG_OTU rows
  select(-cdv_marker_id, -alt_pos, -mult_positions, -amplicon, -cdv_map_feature.name, -gene, -`candidate/random`, -Genaissance_daa_id, -Sequenom_daa_id, -count_amplicons, -count_cmf, -count_gene) #remove unneeded columns originally from the snps dataset 
```

Split the file into two files, one for each group (maize and teosinte), then remove the group information which won't be needed anymore:
```{r}
# initialize the maize and teosinte data files with information that should be in both:
maize.data <- select(joined.data, SNP_ID, Chromosome, Position)
teosinte.data <- select(joined.data, SNP_ID, Chromosome, Position)

# loop through the columns to select those in each group
for (i in 1:length(colnames(joined.data))) {
  stopifnot(joined.data[1,1]=="Group") #check that the first row of joined.data really does contain the "Group" before beginning
  
  # based on the value in the "Group" row, add the column to either the maize or teosinte data set:
  if (joined.data[1,i] %in% c("ZMMIL", "ZMMLR", "ZMMMR")) {
    maize.data <- cbind(maize.data, joined.data[,i])
  }
  if(joined.data[1,i] %in% c("ZMPBA", "ZMPIL", "ZMPJA")) {
    teosinte.data <- cbind(teosinte.data, joined.data[,i])
  }
}

#remove the group information, and format as tibbles again:
maize.data <- as.tibble(filter(maize.data, !(SNP_ID=="Group")))
teosinte.data <- as.tibble(filter(teosinte.data, !(SNP_ID=="Group")))

# check the resulting files:
head(maize.data)
head(teosinte.data)
```

The joined snp and genotype data has been split by group successfully. Now, begin creating output files requested.

First, for each of teosinte and maize, make 10 files (one for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by ?. Missing data is already encoded by ?, however, so this need not be changed. 

```{r}
# first, remove SNPs with multiple or unkonwn positions
maize.known.snps <- filter(maize.data, !(Position %in% c("multiple", "unknown"))) %>%
  filter(!(Chromosome %in% c("multiple", "unknown")))
teosinte.known.snps <- filter(teosinte.data, !(Position %in% c("multiple", "unknown"))) %>%
    filter(!(Chromosome %in% c("multiple", "unknown")))

# Now that there are only integer values in "Chromosome" and "Position" columns, set these as integer rather than character columns
maize.known.snps$Position <- as.integer(maize.known.snps$Position)
maize.known.snps$Chromosome <- as.integer(maize.known.snps$Chromosome)
teosinte.known.snps$Position <- as.integer(teosinte.known.snps$Position)
teosinte.known.snps$Chromosome <- as.integer(teosinte.known.snps$Chromosome)

# Loop through the chromosomes to make output files
for(i in c(1:10)) {
  
  filter(maize.known.snps, Chromosome==i) %>% #filter maize data by chromosome
    dplyr::arrange(Position) %>% # sort by position in increasing order
    write_tsv(paste("../output/Chr", i, ".maize.increasing.txt", sep="")) #write output to file
  
  filter(teosinte.known.snps, Chromosome==i) %>% #repeat above process for teosinte
    dplyr::arrange(Position) %>%
    write_tsv(paste("../output/Chr", i, ".teosinte.increasing.txt", sep=""))
}
```

Next, for each of teosinte and maize, make 10 files (one for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by -. 

```{r}
# As above, using files with only snps of known position, loop through the chromosomes to make output files
for(i in c(1:10)) {
  
  filter(maize.known.snps, Chromosome==i) %>% #filter maize data by chromosome
    dplyr::arrange(desc(Position)) %>% # sort by position in decreasing order
    mutate_all(funs(str_replace_all(., fixed("?"), "-"))) %>% # replace ? with - to encode missing data; ? is a special character, though, so it is escaped using `fixed`
    write_tsv(paste("../output/Chr", i, ".maize.decreasing.txt", sep="")) #write output to file
  
  filter(teosinte.known.snps, Chromosome==i) %>% #repeat above process for teosinte
    dplyr::arrange(desc(Position)) %>%
    mutate_all(funs(str_replace_all(., fixed("?"), "-"))) %>% 
    write_tsv(paste("../output/Chr", i, ".teosinte.decreasing.txt", sep=""))
}
```


## Part II

Visualize the data in this part.

### SNPs per chromosome

Goal: Plot the number of SNPs (only those that are variable WITHIN a given group) by chromosome. Use this figure to determine which group contributes the most SNPs to the dataset that vary intra-group.

To show this, I first need to find the SNPs from `genotypes` that vary within a group, by chromosome. Do this by splitting the genotype file by group, finding which SNPs vary within each group and saving these,  constructing a tibble from this information, and joining this with `snps` to find the chromosome values for each snp:

```{r}
group.snps <- genotypes %>%
  mutate_all(funs(str_replace_all(., fixed("?/?"), NA))) %>% # first recode "?/?"" as NA
  group_by(Group) #group by Group
split.group.snps <- split(group.snps, group.snps$Group) # split dataframes by group and store them in a list

loop.output <- tibble(Group=character(), SNP_ID=character()) # initialize a tibble to store SNPs
for (i in 1:length(split.group.snps)) { # loop through the groups
  for (j in 4:ncol(split.group.snps[[i]])) { # loop through the SNPs
    if (nrow(unique(na.omit(split.group.snps[[i]][,j])))>1) { # look for SNPs with more than one non-NA value
    loop.output <- add_row(loop.output, Group=names(split.group.snps)[i], 
            SNP_ID=colnames(split.group.snps[[i]])[j])
      # save these SNPs (and the group they belong to) to the output tibble
    }
  }
}

# Check that the tibble was filled without any NA values
stopifnot(sum(is.na(loop.output$Group))==0) 
stopifnot(sum(is.na(loop.output$SNP_ID))==0)

# Now, join with `snps` to find the chromosome associated with each SNP:
graph1 <- left_join(loop.output, snps, by="SNP_ID") %>%
  select(Group, SNP_ID, Chromosome) %>% # select only the needed columns
  filter(Chromosome %in% c(1:10)) # remove those SNPs that are on multiple or unknown chromosomes
graph1$Chromosome <- as.numeric(graph1$Chromosome) # format Chromosome as numeric so that it will be sorted numerically rather than alphabetically

```

Now that I have a tibble that contains columns of the group, SNP ID, and chromosome location for each SNP that varies within-group, I can make a bar chart to show what proportion of the number of these SNPs on each chromosome is represented by each group:

```{r}
# Graph the SNPs by chromosome, filling by Group:
ggplot(data=graph1) +
  geom_bar(mapping=aes(x=Chromosome, fill=Group)) +
  scale_x_continuous(breaks=c(1:10), labels=c(1:10)) #set axis labels

# However, this makes it difficult to see proportions contributed by each group; standardize bar heights below:
ggplot(data=filter(graph1, Chromosome %in% c(1:10))) + # filter out the SNPs on unknown or multiple chromosomes
  geom_bar(mapping=aes(x=Chromosome, fill=Group), position="fill") + # standardize bar height
  scale_x_continuous(breaks=c(1:10), labels=c(1:10))

```

The first of these graphs shows that some chromosomes contain more SNPs (defined as those that vary within a group) than others; for example, chromosome 1 contains almost 1500 SNPs but chromosome 10 contains only just over 500 SNPs. The second graph, though, better shows the proportions contributed by each group. In general, each group seems to contribute a fairly consistent proportion of the SNPs across each chromosome; for example, ZMMIL always contributes more than ZDIPL to a given chromosome's SNP count.

### Missing data and amount of heterozygosity

Goal: Make a new column to re-code each SNP as NA (missing data, ?/?), homozygous (A/A, C/C, G/G, or T/T), or heterozygous (otherwise). Then, sort by Group and Species and use to make two graphs to show the proportion of homozygous and heterozygous sites as well as missing data for each group in one graph and for each species in the other graph. In the group graph, normalize the height of the individual bars.

To accomplish this, I do not need the SNPs' positions, only their genotyped values. Therefore, I will use the original `genotypes` file because it is already in an easier layout to work with, with group and species (called `Sample_ID`) already as column names. First, I need to edit the data until it is in "long" form, with three columns: the first with species names, the second with group names, the third with the SNP name, and the fourth with genotypes at that SNP for that species coded as "heterozygous", "homozygous", or NA.

```{r}
# Re-code the genotype dataset:
geno.recode <- genotypes %>%
  mutate_all(funs(str_replace_all(., fixed("?/?"), NA))) %>% #recode all missing data with NA
  mutate_all(funs(str_replace_all(., (fixed("A/A")), "homozygous"))) %>% #recode homozygous pairs
  mutate_all(funs(str_replace_all(., (fixed("C/C")), "homozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("T/T")), "homozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("G/G")), "homozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("A/C")), "heterozygous"))) %>% #recode heterozygous pairs
  mutate_all(funs(str_replace_all(., (fixed("A/G")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("A/T")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("C/A")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("C/G")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("C/T")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("G/A")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("G/C")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("G/T")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("T/A")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("T/C")), "heterozygous"))) %>%
  mutate_all(funs(str_replace_all(., (fixed("T/G")), "heterozygous")))

# Reformat the dataset to tidy, "long" form:
geno.recode.tidy <- geno.recode %>%
  rename(Species=Sample_ID) %>% # rename the "Sample_ID" column to show that it contains species
  select(-JG_OTU) %>% # remove unneeded column
  gather(SNP_name, genotype, -Species, -Group)

```

Next, I can make the two graphs; first make the graph that shows heterozygosity, etc. by species:

```{r}
species.graph2 <- geno.recode.tidy %>%
  arrange(Species) # sort by species

# Make a graph of SNP count by species that is filled with the genotype value
ggplot(data=species.graph2) +
  geom_bar(mapping=aes(x=Species, fill=genotype))
```

As mentioned in the assignment, it is not possible to see species names. However, it is apparent that most species have most sites homozygous, though some have more heterozygous sites than others. In addition, missing data is generally fairly low for most species, but there are a few species with substantially higher missing data counts than the others.

Now, repeat this process with groups instead of species:

```{r, fig.width=10, fig.height=10}
group.graph2 <- geno.recode.tidy %>%
  arrange(Species) # sort by species

# Make a graph of SNP count by species that is filled with the genotype value
ggplot(data=group.graph2) +
  geom_bar(mapping=aes(x=Group, fill=genotype))

# Now, normalize the heights of the individual bars:
ggplot(data=group.graph2) +
  geom_bar(mapping=aes(x=Group, fill=genotype), position="fill")

```

These graphs show the importance of normalizing to compare proportions across groups. In the first graph, without normalizing, the most obvious point is that different groups have drastically different counts of SNPs associated with them, and it is only secondarily apparent that homozygous SNPs appear to be the most common in each group, with lesser numbers of homozygous and missing SNPs. The second graph, with normalization of the bars' heights, shows that some groups have much more missing data than others (e.g., TRIPS has more than 25% missing data for its SNPs but ZMPJA has much less) and some have much higher levels of heterozygosity (e.g., almost 25% of SNPs in ZMPBA are heterozygous, but very few SNPs in ZMMIL are heterozygous).

### My own visualization: A vs C content in basal and derived groups

In 2015, Li et al. published a paper investigating evolutionary patterns of DNA base composition, and in one of their figures compared the A and C content at SNP sites for related species. They found that "derived" species or samples tended to have higher A and lower C (and therefore by Chargaff's rules also higher T and lower G) content than samples that belonged to the ancestral group (which they called "basal"). They hypothesized that this could be due to increased fixation of mutations in a derived population that experienced a bottleneck combined with a known mutation bias towards A:T. (Li, et al. 2015 Nucleic Acids Res. 43(7):3614-3625) 

Goal: Check whether this dataset shows a similar pattern of A vs. C proportions when comparing samples from basal and derived groups (in this case, teosinte and maize).

To accomplish this, first calculate the A and C proportions at all SNP sites for each species, then graph the results by group:

```{r}
AC.prop <- genotypes %>%
  mutate_all(funs(str_replace_all(., fixed("?/?"), ""))) %>% #recode all missing data as "" to make easier to use with `unite()` below
  mutate_all(funs(str_remove_all(., fixed("/")))) %>% # remove all "/" characters
  unite(united, -c(1:3), sep="") %>% # use `unite()` to join all SNPs into one string that can be searched
  mutate(maize.or.teosinte=case_when(Group %in% c("ZMMIL", "ZMMLR", "ZMMMR") ~ "maize",
                                     Group %in% c("ZMPBA", "ZMPIL", "ZMPJA") ~ "teosinte",
                                     TRUE ~ "other"), #make a column to show maize or teosinte
         count.A=str_count(united, "A"), # count each nucleotide in the SNPs
         count.C=str_count(united,"C"),
         count.G=str_count(united, "G"),
         count.T=str_count(united,"T"),
         total=count.A + count.C + count.G + count.T, # find total bp examined
         proportion.A=count.A/total, # find proportions of each NT
         proportion.C=count.C/total,
         proportion.G=count.G/total,
         proportion.T=count.T/total,
         check = proportion.A + proportion.C + proportion.G + proportion.T-1) # make the check row show 0 if all proportions add to 1 as expected
  
    stopifnot(near(0,sum(AC.prop$check))) # make sure that the check row has all 0s as expected


# Graph the proportions by group:
ggplot(AC.prop, aes(proportion.A, proportion.C, color=maize.or.teosinte, shape=maize.or.teosinte)) +
  geom_point() +
  theme(legend.title = element_blank()) +
  xlab("Proportion of A alleles") +
  ylab("Proportion of C alleles")

```

The figure above shows that maize individuals overall do seem to have a somewhat higher proportion of A alleles at any given C allele proportion compared to the teosinte individuals at the SNPs in this dataset, but there is considerable overlap. Interestingly, the "other" individuals, which the UNIX assignment states are Tripsacum individuals, show a more substantial shift toward C allele proportion at a given A allele proportion than either maize or teosinte. Because Tripsacum is in a sense basal to both maize and teosinte, this agrees with the results of the Li, et al. paper. However, if this pattern of allele frequency is in fact related to bottleneck events, then more information about past population size and bottlenecks would be necessary to determine if this dataset does really follow the pattern of A:T enrichment in derived, bottlenecked populations.